File: golang/src/vaani/codility/battleship/Battleship.go
package battleship

import (
	//"fmt"
	//"container/list"
	"strings"
	"strconv"
)

type LinesOfText [][]byte     // A slice of byte slices.

func Build(dimension int) [][]byte {
	board := make([][]byte, dimension)// [dimension][dimension]byte

	for i := 0; i < dimension; i++ {
		board[i] = make([]byte, dimension)
		for j := 0; j < dimension; j++ {
			board[i][j] = 'W';
		}
		//return board
	}
	return board
}

func parseCoordinates(coordinate string) []byte {
	coords := make([]byte, 2)
	coords[0] = coordinate[0] - 49
	coords[1] = coordinate[1] - 65
	//coords[0] = coords[0] - 1
	//fmt.Println(coords)

	return coords;
}

func markShipCoordinates(coordinates string, board [][]byte) {
	tokenizer := strings.Split(coordinates, " ")
	startCoord := tokenizer[0]
	endCoord := tokenizer[1]

	startCoords := parseCoordinates(startCoord);
	startRow := startCoords[0];
	startCol := startCoords[1];

	endCoords := parseCoordinates(endCoord);
	endRow := endCoords[0];
	endCol := endCoords[1];

	for i := startRow; i <= endRow; i++ {
		for j := startCol; j <= endCol; j++ {
			board[i][j] = 'S';
		}
	}
}

func markHit(hit string, board[][] byte) {
	coords := parseCoordinates(hit);
	board[coords[0]][coords[1]] = 'H';
}
func Solution(N int, S string, T string) string {
	sunkShips := 0;
	totalHits := 0;

	// Builds and initializes the board
	board := Build(N);

	//fmt.Println(sunkShips)
	//fmt.Println(totalHits)
	//fmt.Println(board)



	coordinates := strings.Split(S, ",")

	for i := 0; i < len(coordinates); i++ {
		markShipCoordinates(coordinates[i], board);
	}
	//fmt.Println(coordinates)

	hitToken := strings.Split(T, " ")
	//fmt.Println(hitToken)
	for i := 0; i < len(hitToken); i++ {
		hit := hitToken[i]
		//parseCoordinates(hit)
		markHit(hit, board);
	}
	//fmt.Println(board)


	//	// Count hits and sunken ships
	for i := 0; i < len(coordinates); i++ {
		ship := coordinates[i]
		shipTokenizer := strings.Split(ship, " ")
		startCoord := shipTokenizer[0]
		endCoord := shipTokenizer[1]

		startCoords := parseCoordinates(startCoord)
		startRow := startCoords[0]
		startCol := startCoords[1]

		endCoords := parseCoordinates(endCoord)
		endRow := endCoords[0]
		endCol := endCoords[1]
		shipHits := 0
		intactShipParts := 0

		for i := startRow; i <= endRow; i++ {
			for j := startCol; j <= endCol; j++ {
				if (board[i][j] == 'S') {
					intactShipParts++;
				} else if (board[i][j] == 'H') {
					shipHits++;
				}
			}
		}

		if (shipHits > 0) {
			if (intactShipParts == 0) {
				sunkShips++;
			} else {
				totalHits += shipHits;
			}
		}
	}
	return strconv.Itoa(sunkShips) + "," + strconv.Itoa(totalHits)
}

File: src/com/vaani/algo/compete/codility/Battleship.java
package com.vaani.algo.compete.codility;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.StringTokenizer;

public class Battleship {
    public static void main(String[] args) {
        Battleship battleship = new Battleship();
//        String result = battleship.solution(5, "2A 2B,1D 4D", "2A 2B 1D 2D 3D 4D");
//        System.out.println(result);
        
        String result = battleship.solution(4, "1B 2C,2D 4D", "2B 2D 3D 4D 4A");
        System.out.println(result);
    }
    public String solution(int dimension, String ships, String hits) {

        int sunkShips = 0;
        int totalHits = 0;

        // Builds and initializes the board
        char[][] board = build(dimension);
        System.out.println(Arrays.deepToString(board));

        // Parses the ships positions and retrieve the coordinates
        List<String> coordinates = new LinkedList<>();
        
        StringTokenizer tokenizer = new StringTokenizer(ships, ",");
        while(tokenizer.hasMoreTokens()) {
            String shipCoordinates = tokenizer.nextToken();
            markShipCoordinates(shipCoordinates, board);
            coordinates.add(shipCoordinates);
        }
        System.out.println(coordinates);

        // Parses the hits
        StringTokenizer hitToken = new StringTokenizer(hits);
        
        while(hitToken.hasMoreTokens()) {
            String hit = hitToken.nextToken();
//            System.out.println(hit);
            markHit(hit, board);
        }
        
        for(int i=0; i < board.length; i++){
        	System.out.println(Arrays.toString(board[i]));
        }
        // Count hits and sunken ships
        for(String ship : coordinates) {
            StringTokenizer shipTokenizer = new StringTokenizer(ship);

            String startCoord = shipTokenizer.nextToken();
            String endCoord = shipTokenizer.nextToken();

            int[] startCoords = parseCoordinates(startCoord);
            int startRow = startCoords[0];
            int startCol = startCoords[1];

            int[] endCoords = parseCoordinates(endCoord);
            int endRow = endCoords[0];
            int endCol = endCoords[1];

            int shipHits = 0;
            int intactShipParts = 0;
            for(int i = startRow; i <= endRow; i++) {
                for(int j = startCol; j <= endCol; j++) {
                    if(board[i][j] == 'S') {
                        intactShipParts++;
                    } else if(board[i][j] == 'H') {
                        shipHits++;
                    }
                }
            }

            if(shipHits > 0) {
                if(intactShipParts == 0) {
                    sunkShips++;
                } else {
                    totalHits += shipHits;
                }
            }
        }

        return String.format("%d,%d", sunkShips, totalHits);
    }

    private char[][] build(int dimension) {
        char[][] board = new char[dimension][dimension];

        for(int i = 0; i < dimension; i++) {
            for(int j = 0; j < dimension; j++) {
                board[i][j] = 'W';
            }
        }

        return board;
    }

    private int convertLetterToPosition(char position) {
        return (int)Character.toUpperCase(position) - 65;
    }

    private void markShipCoordinates(String coordinates, char[][] board) {
        StringTokenizer tokenizer = new StringTokenizer(coordinates);

        String startCoord = tokenizer.nextToken();
        String endCoord = tokenizer.nextToken();

        int[] startCoords = parseCoordinates(startCoord);
        int startRow = startCoords[0];
        int startCol = startCoords[1];

        int[] endCoords = parseCoordinates(endCoord);
        int endRow = endCoords[0];
        int endCol = endCoords[1];

        for(int i = startRow; i <= endRow; i++) {
            for(int j = startCol; j <= endCol; j++) {
                board[i][j] = 'S';
            }
        }
    }

    private void markHit(String hit, char[][] board) {
        int[] coords = parseCoordinates(hit);
        board[coords[0]][coords[1]] = 'H';
    }

    private int[] parseCoordinates(String coordinate) {
        int[] coords = new int[2];
        coords[0] = Character.getNumericValue(coordinate.charAt(0)) - 1;
        coords[1] = convertLetterToPosition(coordinate.charAt(1));
//        System.out.println(coordinate);
//        System.out.println(Arrays.toString(coords));
        return coords;
    }
}